rspirv/dr/build/autogen_type.rs

```diff
@@ -3,86 +3,130 @@
 // DO NOT MODIFY!
 
 impl Builder {
-    #[doc = "Appends an OpTypeVoid instruction and returns the result id."]
+    #[doc = "Appends an OpTypeVoid instruction and returns the result id, or return the existing id if the instruction was already present."]
     pub fn type_void(&mut self) -> spirv::Word {
+        let mut inst = dr::Instruction::new(spirv::Op::TypeVoid, None, None, vec![]);
+        for ty in &self.module.types_global_values {
```

Comment:
This PR essentially introduces de-duplication logic to DR. Is this desired? It adds a linear search over existing types, which makes the run-time complexity rather unfortunate.
Given that DR's goal is to originally be able to quickly construct a SPIR-V module, this de-duplication seems to be higher level and not really necessary. Something like an optimization pass (spirv-opt?) could take care of this automatically, and probably also faster than running this code on each new type.

Reply:
I'm not sure if spirv-opt can do it, because it's invalid to have spirv with duplicate non-aggregate types.

https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#TypesVariables

I added this because otherwise this needs to end up in userland for everybody building SPIR-V modules, and because it's a pretty small search typically.

Reply:
One option would be to add `type_void_dedup` helpers.